numpy/core/_methods.py

```diff
@@ -201,25 +206,46 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
     # Compute degrees of freedom and make sure it is not negative.
     rcount = max([rcount - ddof, 0])
 
-    # divide by degrees of freedom
+    # Divide by degrees of freedom.
     if isinstance(ret, mu.ndarray):
         ret = um.true_divide(
                 ret, rcount, out=ret, casting='unsafe', subok=False)
+        if is_float16_result and out is None:
+            ret = arr.dtype.type(ret)
     elif hasattr(ret, 'dtype'):
-        ret = ret.dtype.type(ret / rcount)
+        if is_float16_result:
+            ret = arr.dtype.type(ret / rcount)
+        else:
+            ret = ret.dtype.type(ret / rcount)
     else:
         ret = ret / rcount
 
     return ret
 
 def _std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    arr = asanyarray(a)
+
+    is_float16_result = False
     ret = _var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
                keepdims=keepdims)
 
+    # Cast bool, unsigned int, and int to float64 by default
+    if dtype is None:
+        if issubclass(arr.dtype.type, (nt.integer, nt.bool_)):
+            dtype = mu.dtype('f8')
+        elif issubclass(arr.dtype.type, nt.float16):
+            dtype = mu.dtype('f4')
+            is_float16_result = True
```

Comment:
This `if dtype is None` block should probably go before the call to `_var`, so that we supply a dtype to `_var`. Otherwise, `_var` will return a `float16` result, which will lose some precision in the sqrt below.

(The only point of changing `_std` at all is to prevent the tiny amount of precision loss in `sqrt`. Note that if we left `_std` alone, the call to `_var` would return the correct float16 result computed with increased precision, and the unmodified code below would preserve the float16 return type.)